azure-core/azure/core/pipeline/policies/retry.py

```diff
@@ -0,0 +1,251 @@
+# --------------------------------------------------------------------------
+#
+# Copyright (c) Microsoft Corporation. All rights reserved.
+#
+# The MIT License (MIT)
+#
+# Permission is hereby granted, free of charge, to any person obtaining a copy
+# of this software and associated documentation files (the ""Software""), to
+# deal in the Software without restriction, including without limitation the
+# rights to use, copy, modify, merge, publish, distribute, sublicense, and/or
+# sell copies of the Software, and to permit persons to whom the Software is
+# furnished to do so, subject to the following conditions:
+#
+# The above copyright notice and this permission notice shall be included in
+# all copies or substantial portions of the Software.
+#
+# THE SOFTWARE IS PROVIDED *AS IS*, WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+# IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+# FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+# AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+# LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING
+# FROM, OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS
+# IN THE SOFTWARE.
+#
+# --------------------------------------------------------------------------
+"""
+This module is the requests implementation of Pipeline ABC
+"""
+from __future__ import absolute_import  # we have a "requests" module that conflicts with "requests" on Py2.7
+import contextlib
+import logging
+import time
+import re
+from typing import TYPE_CHECKING, List, Callable, Iterator, Any, Union, Dict, Optional  # pylint: disable=unused-import
+import warnings
+
+from azure.core.exceptions import (
+    ServiceRequestError,
+    ConnectionError,
+    ConnectionReadError,
+    raise_with_traceback
+)
+
+from .base import HTTPPolicy, RequestHistory
+
+
+_LOGGER = logging.getLogger(__name__)
+
+
+class RetryPolicy(HTTPPolicy):
+
+    #: Maximum backoff time.
+    BACKOFF_MAX = 120
+
+    def __init__(self, **kwargs):
+        self.total_retries = kwargs.pop('retry_total', 10)
+        self.connect_retries = kwargs.pop('retry_connect', 3)
+        self.read_retries = kwargs.pop('retry_read', 3)
+        self.status_retries = kwargs.pop('retry_status', 3)
+        self.backoff_factor = kwargs.pop('retry_backoff_factor', 0.8)
+        self.backoff_max = kwargs.pop('retry_backoff_max', self.BACKOFF_MAX)
+
+        safe_codes = [i for i in range(500) if i != 408] + [501, 505]
+        retry_codes = [i for i in range(999) if i not in safe_codes]
+        status_codes = kwargs.pop('retry_on_status_codes', [])
+        self._retry_on_status_codes = set(status_codes + retry_codes)
```

Comment:
This looks a bit odd to me. I would expect to retry on:

- Anything that includes a Retry-After header (regardless of http verb and status code)
- 429, 500, 502, 503, 504  (regardless of http verb)
- Write/read timeouts for safe/idempotent methods.

Reply:
Yeah I pasted this from the earlier retry policy - will revise based on common HTTP behaviour.

Reply:
To me, this still looks odd. `self._retry_on_status_codes` includes hundreds of unspecified codes but not 429. Also, the `retry_on_status_codes` kwarg is strictly augmented, making it impossible to exactly configure retryable codes.